### PR TITLE
Limits

### PR DESCRIPTION
**Pull request by shargon**
I've tried to eliminate as many peeks as possible,

Note that the GetPrice funtion are `protected` and now have one parameter

```protected virtual long GetPrice(OpCode nextInstruction)```

This class are not extended, in neo/neo-vm/neo-cli/neo-gui

### CODE REVIEW

File: neo/SmartContract/StorageContext.cs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 namespace Neo.SmartContract
 {
-    internal class StorageContext : IInteropInterface
+    public class StorageContext : IInteropInterface
```
Comment (erikzhang) [inline neo/SmartContract/StorageContext.cs]: Please do not change this modifier.

File: neo/SmartContract/ApplicationEngine.cs [hunk 2]
```diff
@@ -100,17 +112,104 @@ private bool CheckItemSize()
...
+        /// <summary>
+        /// Check if the BigInteger is allowed for numeric operations
+        /// </summary> 
+        private bool CheckBigIntegers(OpCode nextInstruction)
+        {
+            switch (nextInstruction)
+            {
+                case OpCode.INC:
+                    {
+                        BigInteger x = EvaluationStack.Pop().GetBigInteger();
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]: Should use `Peek()`

Comment (shargon) [reply neo/SmartContract/ApplicationEngine.cs]: Yes xD

File: neo/SmartContract/ApplicationEngine.cs [hunk 3]
```diff
@@ -100,17 +112,104 @@ private bool CheckItemSize()
...
+                            return false;
+
+                        break;
+                    }
+                case OpCode.MUL:
+                    {
+                        BigInteger x2 = EvaluationStack.Peek().GetBigInteger();
+                        BigInteger x1 = EvaluationStack.Peek(1).GetBigInteger();
+
+                        if (!CheckBigInteger(x2) || !CheckBigInteger(x1) || !CheckBigInteger(x1 * x2))
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]: I think this check will consume as much resources as normal execution.

Comment (shargon) [reply neo/SmartContract/ApplicationEngine.cs]: Maybe the operation does not have to be verified?

Comment (shargon) [reply neo/SmartContract/ApplicationEngine.cs]: Or maybe in neo-vm could create virtual method like 
```protected virtual bool CheckMulOpCode(BigItenger a,BigInteger b);```

Comment (erikzhang) [reply neo/SmartContract/ApplicationEngine.cs]: Can we check the length of `x1` and `x2`, and then predict the length of result?

File: neo/SmartContract/ApplicationEngine.cs [hunk 4]
```diff
@@ -100,17 +112,111 @@ private bool CheckItemSize()
...
+                            return false;
+
+                        break;
+                    }
+                case OpCode.DIV:
+                    {
+                        BigInteger x2 = EvaluationStack.Peek().GetBigInteger();
+                        BigInteger x1 = EvaluationStack.Peek(1).GetBigInteger();
+
+                        if (!CheckBigInteger(x2) || !CheckBigInteger(x1) || !CheckBigInteger(x1 / x2))
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]: If `CheckBigInteger(x2) == true` and `CheckBigInteger(x1) == true`, then it is impossible that `CheckBigInteger(x1 / x2)) == false`.

Comment (shargon) [reply neo/SmartContract/ApplicationEngine.cs]: You are right!

File: neo/SmartContract/ApplicationEngine.cs [hunk 5]
```diff
@@ -143,7 +143,7 @@ private bool CheckBigIntegers(OpCode nextInstruction)
                     {
                         BigInteger x = EvaluationStack.Peek().GetBigInteger();
 
-                        if (!CheckBigInteger(x) || !CheckBigInteger(x - 1))
+                        if (!CheckBigInteger(x))
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]: If `x` is a negative number......

File: neo/SmartContract/ApplicationEngine.cs [hunk 6]
```diff
@@ -143,7 +143,7 @@ private bool CheckBigIntegers(OpCode nextInstruction)
                     {
                         BigInteger x = EvaluationStack.Peek().GetBigInteger();
 
-                        if (!CheckBigInteger(x))
+                        if (!CheckBigInteger(x) || ( x <= 0 && !CheckBigInteger(x - 1)))
```
Comment (erikzhang) [inline neo/SmartContract/ApplicationEngine.cs]: `x.Sign <= 0` is better then `x <= 0`

**Comment by shargon** [comment]
What do you think about move the restrictions
```
if (CurrentContext.InstructionPointer < CurrentContext.Script.Length)
                    {
                        OpCode nextOpcode = CurrentContext.NextInstruction;

                        gas_consumed = checked(gas_consumed + GetPrice(nextOpcode) * ratio);
                        if (!testMode && gas_consumed > gas_amount) return false;

                        if (!CheckItemSize(nextOpcode)) return false;
                        if (!CheckStackSize(nextOpcode)) return false;
                        if (!CheckArraySize(nextOpcode)) return false;
                        if (!CheckInvocationStack(nextOpcode)) return false;
                        if (!CheckBigIntegers(nextOpcode)) return false;
                    }
```

Inside the `StepInto` at ApplicationEngine
```
        public void StepInto()
        {
            if (InvocationStack.Count == 0) State |= VMState.HALT;
            if (State.HasFlag(VMState.HALT) || State.HasFlag(VMState.FAULT)) return;
            OpCode opcode = CurrentContext.InstructionPointer >= CurrentContext.Script.Length ? OpCode.RET : (OpCode)CurrentContext.OpReader.ReadByte();
            try
            {
                ExecuteOp(opcode, CurrentContext);
            }
            catch
            {
                State |= VMState.FAULT;
            }
        }
```

When my debugger use StepInto method, i have no real limits, and i have no way to call the same method only once, this could save the line 
``` OpCode nextOpcode = CurrentContext.NextInstruction;```

because `StepInto` already have this variable, only need change the state to `FAULT` if this checks return `false`

**Comment by erikzhang** [comment]
I don't think it's a good idea. Because some other blockchains are also using NeoVM for smart contract. We should not add the restrictions on NEO to the VM.

**Comment by erikzhang** [comment]
@shargon If it is ready, I will merge this PR.

**Comment by shargon** [comment]
Let me do the last tests this afternoon and I confirm you today (Spain time)

**Comment by shargon** [comment]
This try/catch is already in Execute method. And haven't got generic catch. Otherwise, I think it's OK.

**Pull request merged**